xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-realtime/plugin.js

```diff
@@ -47,16 +47,198 @@
+        }
+
+      } else if (editor.mode === 'source' && newMode === 'wysiwyg') {
+        // Swithing from source  to wysiwyg mode.
+        const realtimeCheckbox = editor._realtimeInterface.getAllowRealtimeCheckbox();
+
+        // We only need to change the behavior if we were in a realtime session before switching to source.
+        if (editor._realtimeSource.realtime) {
+
+          // Before switching to wysiwyg, check wether the source was edited.
+          const sourceDirty = editor._realtimeSource.previousValue !== editor.getSnapshot();
+
+          // There are unsaved changes if there were unsaved changes before switching to source
+          // or if there were changes made while we were in source view.
+          const dirty = editor._realtimeSource.dirty || sourceDirty;
+
+          const dataReady = function () {
+            // After switching to wysiwyg.
+
+            // Bulletproofing, in iframe mode, when switching to wysiwyg, setData is called multiple times.
+            if (editor.mode != 'wysiwyg') {
+              editor.once('dataReady', dataReady);
+              return;
+            }
+
+            // Update the realtime channels to prepare joining the realtime session,
+            // as well as knowing if there are users in the session.
+            editor._realtime._updateChannels().then(() => {
+
+              // When the editor is dirty, we can join only if we are alone.
+              if (dirty) {
+                /*jshint -W106 */
+                if (editor._realtime._realtimeContext.channels.wysiwyg_users > 0) {
+                  /*jshint +W106 */
+
+                  // Bring the autosave checkbox back.
+                  editor._realtimeInterface.realtimeAllowed(false);
+
+                  // Show a notification explaining that we are not rejoining the realtime session.
+                  editor.showNotification(
+                    editor.localization.get('xwiki-realtime.notification.sourcearea.notRejoiningSession'),
+                    'warning',
+                    5000);
+                } else {
+                  // Join the realtime session.
+
+                  // Show a notification explaining that we are rejoining the session because we are alone.
+                  editor.showNotification(
+                    editor.localization.get('xwiki-realtime.notification.sourcearea.rejoiningSession.alone'),
+                    'success',
+                    5000);
+                  realtimeCheckbox.prop('checked', true);
+                  editor._realtime._startRealtimeSync();
+                }
+              } else {
+                // Join the realtime session.
+                editor.showNotification(
+                  editor.localization.get('xwiki-realtime.notification.sourcearea.rejoiningSession.noChanges'),
+                  'success',
+                  5000);
+                realtimeCheckbox.prop('checked', true);
+
+                const readOnly = function () {
```

Comment:
Too bad ``_startRealtimeSync()`` is not async / doesn't return a promise. Unfortunately ``ChainPadNetflux.start()`` doesn't provide a simple / clean way to wait for the connection to be ready...